### PR TITLE
Fix LongformerModel hidden states

### PR DESCRIPTION
**Pull request by ydshieh**
# What does this PR do?

`LongformerModel` currently returns `hidden_states` that is padded, while `TFLongformerModel` and `LEDEncoder` return un-padded version.

This PR makes `LongformerModel` return the un-padded `hidden_states` .

----

`LEDEncoder` also returns  un-padded `attentions`. Maybe the same logic should be applied to `LongformerModel` too.

### CODE REVIEW

**Comment by ydshieh** [comment]
The test needs to be adjusted due to the change. But I would like to have some feedback first before proceeding.

File: src/transformers/models/longformer/modeling_longformer.py [hunk 1]
```diff
@@ -1705,17 +1705,26 @@ def forward(
...
         # undo padding
+        all_hidden_states = None
+        if output_hidden_states:
+            all_hidden_states = encoder_outputs[1]
+
         if padding_len > 0:
             # unpad `sequence_output` because the calling function is expecting a length == input_ids.size(1)
             sequence_output = sequence_output[:, :-padding_len]
+            if output_hidden_states:
+               all_hidden_states = tuple([state[:, :-padding_len] for state in all_hidden_states])
```
Comment (patil-suraj) [inline src/transformers/models/longformer/modeling_longformer.py:1716]: Good catch! Will let @patrickvonplaten take final look.

**Comment by patil-suraj** [comment]
@patrickvonplaten could you take a look here ?

**Review by patrickvonplaten — COMMENTED** [review]
Very much agree with you here @ydshieh !

**Comment by ydshieh** [comment]
Thanks for reviewing! I will finalize this PR by adjusting the test.

**Comment by ydshieh** [comment]
@patrickvonplaten @patil-suraj 

A minor question:

# TL;DR

Current  `LongformerModel` / `TFLongformerModel` don't unpad `attentions`. I am going to unpad the `attentions` just like what have been done for `LEDModel` & `TFLEDModel`. 

This is required to make `TFLongformerModelTest` pass (in an easy way).

Does this sound good to you?

# TL;TR

In `LEDModel` & `TFLEDModel`, the `attentions` are also un-padded:

https://github.com/huggingface/transformers/blob/57882177becb85560f1ff931abb1b0b75d67e70d/src/transformers/models/led/modeling_led.py#L1871-L1872

https://github.com/huggingface/transformers/blob/57882177becb85560f1ff931abb1b0b75d67e70d/src/transformers/models/led/modeling_tf_led.py#L1814-L1819



However, `LongformerModel` / `TFLongformerModel` don't perform unpadding for `attentions` (but do unpadding for `hidden_states`). This causes a problem for `TFLongformerModelTester`:

https://github.com/huggingface/transformers/blob/57882177becb85560f1ff931abb1b0b75d67e70d/tests/test_modeling_tf_longformer.py#L77-L81

The testings for both `attentions` and `hidden_states` are controlled by `self.encoder_seq_length` => so we need to unpad `attentions` too (because `hidden_states` are unpadded).

File: src/transformers/models/longformer/modeling_longformer.py [hunk 2]
```diff
@@ -1249,6 +1249,7 @@ def forward(
         output_attentions=False,
         output_hidden_states=False,
         return_dict=True,
+        padding_len=0,
```
Comment (ydshieh) [inline src/transformers/models/longformer/modeling_longformer.py:1252]: I hope adding this new argument is fine.

File: src/transformers/models/longformer/modeling_longformer.py [hunk 3]
```diff
@@ -1700,15 +1711,11 @@ def forward(
...
+            padding_len=padding_len,
         )
         sequence_output = encoder_outputs[0]
         pooled_output = self.pooler(sequence_output) if self.pooler is not None else None
 
-        # undo padding
-        if padding_len > 0:
-            # unpad `sequence_output` because the calling function is expecting a length == input_ids.size(1)
-            sequence_output = sequence_output[:, :-padding_len]
-
```
Comment (ydshieh) [inline src/transformers/models/longformer/modeling_longformer.py:1711]: Move the unpadding logic to `LongformerEncoder`, so we don't need to deal with `encoder_outputs[1:]` or `encoder_outputs[2:]` here (which depends on the combination of `output_hidden_states`/`output_attentions`).

File: src/transformers/models/longformer/modeling_longformer.py [hunk 4]
```diff
@@ -1246,6 +1246,7 @@ def forward(
         hidden_states,
         attention_mask=None,
         head_mask=None,
+        padding_len=0,
```
Comment (ydshieh) [inline src/transformers/models/longformer/modeling_longformer.py:1249]: I hope adding this new argument is fine.

`TFLongformerEncoder` has already this argument.

(and moving the unpadding logic here makes the code a bit easier),  i.e no need to worry `encoder_outputs[1]`,  `encoder_outputs[2]` etc in `LongformerModel`

```
        if not return_dict:
            return (sequence_output, pooled_output) + encoder_outputs[1:]
```

File: src/transformers/models/longformer/modeling_longformer.py [hunk 5]
```diff
@@ -1697,18 +1708,14 @@ def forward(
...
             return_dict=return_dict,
         )
         sequence_output = encoder_outputs[0]
         pooled_output = self.pooler(sequence_output) if self.pooler is not None else None
 
-        # undo padding
-        if padding_len > 0:
-            # unpad `sequence_output` because the calling function is expecting a length == input_ids.size(1)
-            sequence_output = sequence_output[:, :-padding_len]
-
```
Comment (ydshieh) [inline src/transformers/models/longformer/modeling_longformer.py:1711]: Move the unpadding logic to `LongformerEncoder`, so we don't need to deal with `encoder_outputs[1:]` or `encoder_outputs[2:]` here (which depends on the combination of `output_hidden_states`/`output_attentions`).

**Review by patrickvonplaten — APPROVED** [review]
Great refactor!
